### PR TITLE
Fix/faster contains

### PR DESCRIPTION
**Pull request by unfinishedteleporter**
Improves performance in some situations (when finding which geometries contain a single one). Another outcome is a refactoring of the way it works to be easier to followup.

### CODE REVIEW

**Review by johntwyman — APPROVED** [review]

**Review by seamuslee001 — APPROVED** [review]

**Review by unfinishedteleporter — COMMENTED** [review]
There can only ever be 1 or 0 results from the query. They way it is written implies otherwise, and is also a bit unclear because if one of the geoms does not exist, the function will return null rather than false (contradicting doc). This isn't seen in the api because it treats null same as false. I'll change it so its clearer.

File: CRM/CiviGeometry/BAO/Geometry.php [hunk 1]
```diff
@@ -80,65 +80,99 @@ public static function getCentroid($params) {
...
     }
-    if (is_numeric($params['geometry_b'])) {
-      $sql = $dualIntegerSQL . $dualIntegerWhere . " AND a.id = %1";
-      $sql_params = [
-        1 => [$params['geometry_a'], 'Positive'],
-        2 => [$params['geometry_b'], 'Positive'],
+
+    $result = CRM_Core_DAO::executeQuery("SELECT $selectStr FROM $fromStr WHERE $whereStr", $sqlParams);
+    while ($result->fetch()) {
+      return $result->contains_result == '1';
```
[the remark this answers is not in the document]
Comment (unfinishedteleporter) [reply CRM/CiviGeometry/BAO/Geometry.php]: Ok, I belivee 694206b3b3598fa3f3a724c53a3fbe14c75f36bb resolves this issue

**Comment by unfinishedteleporter** [comment]
Pending passing some new tests in buildkite.

File: tests/phpunit/api/v3/GeometryTest.php [hunk 2]
```diff
@@ -271,6 +271,16 @@ public function testMySQLSTContains() {
     $this->callAPISuccess('Geometry', 'delete', ['id' => $upperHouseDistrictMBR['id']]);
     $this->callAPISuccess('GeometryType', 'delete', ['id' => $UHGeometryType['id']]);
     $this->callAPISuccess('GeometryCollection', 'delete', ['id' => $UHCollection['id']]);
+
+    // Check that providing an id for either geometry that doesn't exist fails
+   $this->callAPIFailure('Geometry', 'contains', [
+      'geometry_a' => 99999,
+      'geometry_b' => $upperHouseDistrict['id'],
+    ], 'Geometry 99999 Does not exist in the database');
```
Comment (seamuslee001) [inline tests/phpunit/api/v3/GeometryTest.php:279]: @unfinishedteleporter need to have a # in front of the numbers as per https://github.com/australiangreens/au.org.greens.civigeometry/pull/40/files#diff-e476444ffc64748061579ed65e3c31a76e69a9f0dad6cf541b9f021bd7dbbe8eR380

Comment (unfinishedteleporter) [reply tests/phpunit/api/v3/GeometryTest.php]: Yeah, whoops. I also didn't realise it actually checked that message, thought it was just to display it.

**Comment by unfinishedteleporter** [comment]
The tests pass 🎉

**Pull request merged**